### ISSUE TITLE
Can't remove the element using update

### ISSUE THREAD

**Issue by adamglowacki** [nikita-volkov/focus#5]
Function `update` leaves the element instead of removing it when the passed function returns `Nothing`.

Based on https://github.com/nikita-volkov/stm-containers/issues/28.

**Comment by nikita-volkov** [comment]
Fix is released in 1.0.1.4

**Comment by carbolymer** [comment]
It does not seem to be fixed in 1.0.1.4 (either in focus or stm-containers / hamt). The code below loops infinitely:
```haskell
#!/usr/bin/env stack
{- stack
  script
  --resolver lts-16.26
  --package stm
  --package async
  --package focus
  --package transformers
  --package hashable
  --package unsafe
  --package containers
  --package nonempty-containers
  --package stm-containers
  --extra-dep stm-hamt-1.2.0.4@sha256:7957497c022554b7599e790696d1a3e56359ad99e5da36a251894c626ca1f60a,3970
  --extra-dep primitive-extras-0.8@sha256:fca0310150496867f5b9421fe1541ecda87fae17eae44885a29f9c52dd00c8ff,2963
  --extra-dep primitive-unlifted-0.1.3.0@sha256:a98f827740f5dcf097d885b3a47c32f4462204449620abc9d51b8c4f8619f9e6,1427
-}

import           Control.Concurrent
import           Control.Concurrent.STM
import           Control.Monad
import           Data.Hashable          (Hashable)
import           Data.Maybe
import qualified Data.Sequence          as S
import           Data.Sequence.NonEmpty (NESeq, (|>))
import qualified Data.Sequence.NonEmpty as NES
import qualified Focus                  as F
import qualified StmContainers.Map      as SM
import           System.IO.Unsafe
import           System.Unsafe

popQueue :: (Hashable k, Eq k) => k -> SM.Map k (NESeq v) -> IO (Maybe v)
popQueue key queue = atomically $ do
  element'm <- SM.focus popQueueF key queue
  when (isNothing element'm) $ SM.delete key queue -- this doesn't delete either!!!
  pure element'm
  where
    popQueueF = do
      element'm <- fmap (fmap NES.head) F.lookup
      queue'm <- fmap (join . (fmap (NES.nonEmptySeq . NES.tail))) F.lookup
      unsafePerformIO $ do -- debugging
        let elem_ = coerce element'm :: Maybe (String)
        let elem_t = coerce queue'm :: Maybe (NESeq String)
        putStrLn "***"
        putStrLn $ show $ elem_
        putStrLn $ show $ elem_t
        putStrLn "***"
        threadDelay 100000
        pure F.lookup
      -- F.alter (join . fmap (NES.nonEmptySeq . NES.tail))
      case queue'm of
        Just queue_ -> F.insert queue_
        Nothing     -> F.delete -- doesn't delete!!!
      pure element'm

pushQueue :: (Hashable k, Eq k) => k -> v -> SM.Map k (NESeq v) -> IO ()
pushQueue key value = atomically . SM.focus pushQueueF key
  where
    pushQueueF = F.insertOrMerge (<>) (NES.singleton value)


main :: IO ()
main = do
  putStrLn "\n"
  queue <- SM.newIO :: IO (SM.Map Int (NESeq String))
  pushQueue 1 "foo" queue
  pushQueue 1 "bar" queue
  pushQueue 2 "fooz" queue
  pushQueue 2 "baz" queue
  putStrLn $ "DRAIN-----1"
  drainQ 1 queue putStrLn
  putStrLn $ "DRAIN-----2"
  drainQ 2 queue putStrLn
    where
      drainQ :: (Hashable k, Eq k) => k -> SM.Map k (NESeq v) -> (v -> IO a) -> IO ()
      drainQ key queue act = do
        element'm <- popQueue key queue
        case element'm of
          Just element -> do act element
                             drainQ key queue act
          Nothing -> pure ()
```
